ros2cli/ros2cli/daemon/__init__.py

```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
             server.server_close()
 
 
+class _DirectNode:
+    """A direct node, that resets itself when a network interface changes."""
+
+    def __init__(self, args):
+        self.args = args
+        # TODO(ivanpauno): A race condition is possible here, since it isn't possible to know
+        # exactly which interfaces were available at node creation.
+        self.node = DirectNode(args)
+        self.addresses_at_start = self.interfaces_ip_addresses()
+
+    def __enter__(self):
+        self.node.__enter__()
+        return self
+
+    def __getattr__(self, name):
+        attr = self.node.__getattr__(name)
+
+        def wrapper():
+            self.reset_if_addresses_changed()
+            return self.node.__getattr__(name)()
+        wrapper.__name__ = attr.__name__
+
+        if inspect.ismethod(attr):
+            return wrapper
+        self.reset_if_addresses_changed()
+        return attr
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.node.__exit__(exc_type, exc_value, traceback)
+
+    def interfaces_ip_addresses(self):
+        addresses_by_interfaces = {}
+        for (kind, info) in netifaces.gateways().items():
+            if kind not in (netifaces.AF_INET, netifaces.AF_INET6):
+                continue
+            print('Interface kind: {}, info: {}'.format(kind, info))
+            if isinstance(info, dict):
```

Comment:
@ivanpauno what are we leaving out with this `if` clause?

Reply:
```python3
import netifaces
netifaces.gateways().items()
```

output:
```
dict_items([('default', {2: ('172.17.0.1', 'eth0')}), (2, [('172.17.0.1', 'eth0', True)])])
```

As you can see, the default interface value is a dictionary like `{kind: info_tuple}`,
though the others interface values are a list of `info_tuple`.
I think it would be clearer if I do:
```python3
if 'default' == kind:
    continue
```

do you agree?

Reply:
Also, I'm checking, but I think this other if wasn't needed:
https://github.com/ros2/ros2cli/blob/106b33459f79d2a883999b15f3df3485fadab762/ros2cli/ros2cli/daemon/__init__.py#L138-L139

Reply:
Also, I think I can delete this if https://github.com/ros2/ros2cli/blob/106b33459f79d2a883999b15f3df3485fadab762/ros2cli/ros2cli/daemon/__init__.py#L138-L139.

Will double check.

Reply:
Ok, I simplified the logic.
`'default'` was already ruled out in:
```python3
if kind not in (netifaces.AF_INET, netifaces.AF_INET6):
    continue
```

> Is it guaranteed it'll be the only interface value that is a dict?

Yes. For further reference, see https://pypi.org/project/netifaces/.